### PR TITLE
Fix to distribution.__repr__ with lazy attributes

### PR DESCRIPTION
**Pull request by neerajprad**
`__repr__` currently fails for distributions with lazy attributes in PyTorch master, throwing a `KeyError`. This fixes the issue. 

**Additionally:**
 - Added `logits` to `arg_constraints` for distributions that accept either `probs` or `logits`. This is both to have `__repr__` display the `logits` param when available, and to be able to do validation checks (e.g. NaN checks) when the logit parametrization is used. @fritzo, @alicanb - I think there were reasons why we had not done so in the first place, but I am unable to recall now. It passes all the tests, but let me know if there is something that I am missing at the moment.
 - There are certain distributions, e.g. `OneHotCategorical` which won't show any parameters because it uses a `categorical` instance under the hood and neither `logits` / `probs` in `arg_constraints` are present in the instance's `__dict__`. This isn't addressed in this PR. 

cc. @vishwakftw, @fritzo, @nadavbh12, @apaszke

### CODE REVIEW

File: torch/distributions/distribution.py [hunk 1]
```diff
@@ -221,7 +221,7 @@ def _validate_sample(self, value):
             raise ValueError('The value argument must be within the support')
 
     def __repr__(self):
-        param_names = [k for k, _ in self.arg_constraints.items()]
+        param_names = [k for k, _ in self.arg_constraints.items() if k in self.__dict__]
```
Comment (vishwakftw) [inline torch/distributions/distribution.py]: Nice catch!!. I have one concern though: in MVNs, if we happen to lazily evaluate *all* the parameters (`covariance_matrix`, `scale_tril` and `precision_matrix`), won't they show up redundantly in the `__repr__`?

**Review by apaszke — APPROVED** [review]

File: torch/distributions/distribution.py (hunk 1 above)
Comment (neerajprad) [reply torch/distributions/distribution.py]: You are right that if all the lazy parameters happen to have been evaluated, they will show up in `__repr__` even if they are redundant (also true for classes with logits and probs)! The downside then is that the output we get isn't guaranteed to be static (though there shouldn't be a reason to expect it to be static), but the upside is that it can aid debugging by letting us know which attributes have been set without having to go through the instance's `__dict__`. Since we are only reporting the size, I think it's fine to report any redundant parameters for the distribution instance. Preventing this would likely require more complex logic than is worth (at least not until we have other use cases for more elaborate ways of registering params).

Comment (vishwakftw) [reply torch/distributions/distribution.py]: Thanks for the views @neerajprad . I understand the reason, and yes: shapes really shouldn't be an issue.

**Comment by apaszke** [comment]
Apparently this has been merged, but the bot hasn't closed the PR for some reason.
